core/embed/lib/error_handling.h

```diff
@@ -20,6 +20,167 @@
 #ifndef LIB_ERROR_HANDLING_H
 #define LIB_ERROR_HANDLING_H
 
+// Suppresses the intellisense error in VSCode
+#ifndef __FILE_NAME__
+#define __FILE_NAME__ __FILE__
+#endif
+
+// Status code type
+typedef struct {
+  // Do not access this field directly,
+  // use `ts_ok()` and `ts_error()` macros.
+  uint32_t code;
+} ts_t;
+
+// Build status code from any 16-bit value.
+//
+// Status codes are hardened against fault injections
+// by storing the same value in the upper 16 bits.
+#define TS_BUILD(code) ((ts_t){(code) | ((code) << 16)})
+
+// OK status code (signalling success or no error)
+#define TS_OK TS_BUILD(0)
+
+// This offset ensures at lest 16bit hamming distance
+// between `TS_OK` and other status codes.
+#define TS_ERROR_OFFSET 0xFF00
+
+// Build error status code from any 8-bit value
+//
+// The code should be in range 0 to 255 to ensure
+// 16-bit hamming distance to `TS_OK`
+#define TS_ERROR_BUILD(code) TS_BUILD(TS_ERROR_OFFSET + (code))
+
+// Error status codes
+#define TS_ERROR TS_ERROR_BUILD(0)
+#define TS_ERROR_BUSY TS_ERROR_BUILD(1)
+#define TS_ERROR_TIMEOUT TS_ERROR_BUILD(2)
+#define TS_ERROR_NOTINIT TS_ERROR_BUILD(3)
+#define TS_ERROR_ARG TS_ERROR_BUILD(4)
+#define TS_ERROR_IO TS_ERROR_BUILD(5)
+
+// Check status code consistency and returns its value.
+// If invalid status code is detected, it will call `__fatal_error()`.
+#define _ts_checked(status)                                       \
+  ({                                                              \
+    ts_t _checked = (status);                                     \
+    if ((_checked.code & 0xFFFF) != (_checked.code >> 16)) {      \
+      __fatal_error("ts_check() error", __FILE_NAME__, __LINE__); \
+    }                                                             \
+    _checked;                                                     \
+  })
+
+// Returns `true` if status code is `TS_OK`
+#define ts_ok(status) (_ts_checked(status).code == TS_OK.code)
+
+// Returns `true` if status code is NOT `TS_OK`
+#define ts_error(status) (_ts_checked(status).code != TS_OK.code)
```

Comment:
If there will be multiple error states, then we should also have `ts_eq()`.

Reply:
It might be also nice to introduce a `ts_switch(status)` macro which expands to `switch(_ts_checked(status).code)`.

Reply:
I think the mechanism needs to allow easy translation of error codes and the ability to suppress certain errors or treat them as success.  This should make that possible.

Reply:
I've added `ts_eq()` and `ts_string()`, see 8b027fbcf0760c154cfebb56e5dddf01f4903e33.

The following code cannot be compiled since `TS_xxx` is a structure. For some reason, the compiler cannot deduce that TS_OK.code is a constant expression... :-(.

```c
switch (status.code) {
  case TS_OK.code:
  case TS_ERROR.code:
}
``` 
Maybe we can give up using `case` for error codes and always use if. But I'll try to find some solution.

Reply:
```c
#define ts_case(status) case status.code

ts_switch(f()) {
  ts_case(TS_OK):
  // ...
  break;
default:
  // ...
}

Reply:
You mentioned IRL some issue with the compiler not being able to treat the value as a constant. In C++ defining `TS_OK` etc. using `constexpr` would solve this, but that probably isn't an option in C. Another idea is something like this:
```
#define ts_case(status) case status##__value
#define TS_OK__value TS_BUILD(0)
#define TS_OK ((const ts_t){TS_OK__value})
```